design/ceph/multus-network.md

```diff
@@ -78,7 +78,78 @@ Nothing to do in particular since they don't use any service IPs.
 
 ### CSI pods
 
-We can add annotations to these pods and they can reach out the Ceph public network, then the driver will expose the block or the filesystem normally.
+Steps must be taken to fix a CSI-with-multus issue documented
+[here](https://github.com/rook/rook/issues/8085). To summarize the issue:
+when a CephFS/RBD volume is mounted in a Pod using cephcsi and then the CSI CephFS/RBD plugin is
+restarted or terminated (e.g. by restarting or deleting its DaemonSet), all operations on the volume
+become blocked, even after restarting the CSI pods. The only workaround is to restart the node where
+the cephcsi plugin pod was restarted.
+
+When deploying a CephCluster resource configured to use multus networks, a multus-connected network
+interface will be added to the host network namespace for all nodes that will run CSI plugin pods.
+
+The Rook-Ceph Operator's CSI controller creates a DaemonSet configured to use the
+`network.selectors.public` network specified for the CephCluster. This DaemonSet runs on all the
+nodes that will have CSI pods. Its Pods exist to "hold" a particular network interface that the CSI
+pods can reliably connect to. These Pods will be referred to as the "holder" Pods.
+
+These Pods should only be stopped and restarted when a node is stopped so that volume operations do
+not become blocked. The Rook-Ceph Operator's CSI controller should set the DaemonSet's update
+strategy to `OnDelete` so that the pods do not get deleted if the DaemonSet is updated while also
+ensuring that the pods will be updated on the next node reboot.
+
+Furthermore, a new Rook-specific CSI sidecar will run in CSI plugin Pods alongside the Ceph CSI
+plugin process. This sidecar will watch for holder Pods, and when a holder Pod is created or
+updated, it will move the public network from the Pod's network namespace to the host node's network
+namespace. In this way, even if the CSI pods are restarted, the network interface will not change,
+and the operations to the Ceph volume will not be blocked.
+
+If a holder Pod is deleted, the interface hold will be lost, so the sidecar should remove the
+interface from the host's network namespace.
+
+If the sidecar container is stopped, it should move all holder Pod namespaces from the host's
+network namespace back to the holder Pod's network namespace under the assumption that the CSI
+plugin is being removed, possibly by a CephCluster being deleted or the node going down for
+maintenance.
+TODO: This design will disrupt block/file traffic when CSI pods are restarted. It seems better to
```

Comment:
During the CSI upgrade I thought we could avoid moving the interface and thus keep things working in that case, but good question about how to detect the difference between the node drain and the csi pod update...

Reply:
Once the CSI upgrade is requested we know we should not move the interface back. The good part is that the upgrade should be handled by the Operator, so it knows when the CSI driver is updated. This information (status field somewhere or similar) can be shared so that when the CSI pod received SIGTERM it looks up this information and then determine whether it should move the interface back or not. Obviously, this won't work if this takes longer than the termination grace period, but that should not happen.

Reply:
Ideally, if it was an upgrade, we could detect it and leave the interfaces. One issue is that there are corner cases (even if rare) where someone might start an upgrade but the pod won't be restarted on this node (maybe the node selectors changed and the pod won't be rescheduled on the node, or maybe someone drains the node during upgrade). In that case, we will leave interfaces that might interfere with other applications. Even though un-moving interfaces will cause I/O hiccups, I believe it is safer for the system's long-term stability.